cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -756,211 +1078,227 @@ class StructConverter : public TypedConverterVisitor<StructBuilder, StructConver
-                                    std::shared_ptr<Array>* out) {
+Status ConvertPySequence(PyObject* sequence_source, PyObject* mask,
+                         const PyConversionOptions& options,
+                         std::shared_ptr<ChunkedArray>* out) {
   PyAcquireGIL lock;
 
+  PyDateTime_IMPORT;
+
   PyObject* seq;
   OwnedRef tmp_seq_nanny;
 
   std::shared_ptr<DataType> real_type;
 
-  RETURN_NOT_OK(ConvertToSequenceAndInferSize(obj, &seq, &size));
+  int64_t size = options.size;
+  RETURN_NOT_OK(ConvertToSequenceAndInferSize(sequence_source, &seq, &size));
   tmp_seq_nanny.reset(seq);
-  if (type == nullptr) {
+
+  // In some cases, type inference may be "loose", like strings. If the user
+  // passed pa.string(), then we will error if we encounter any non-UTF8
+  // value. If not, then we will allow the result to be a BinaryArray
+  bool strict_conversions = false;
+
+  if (options.type == nullptr) {
     RETURN_NOT_OK(InferArrowType(seq, &real_type));
   } else {
-    real_type = *type;
+    real_type = options.type;
+    strict_conversions = true;
   }
   DCHECK_GE(size, 0);
 
   // Handle NA / NullType case
   if (real_type->id() == Type::NA) {
-    out->reset(new NullArray(size));
+    ArrayVector chunks = {std::make_shared<NullArray>(size)};
+    *out = std::make_shared<ChunkedArray>(chunks);
     return Status::OK();
   }
 
-  // Give the sequence converter an array builder
-  std::unique_ptr<ArrayBuilder> builder;
-  RETURN_NOT_OK(MakeBuilder(pool, real_type, &builder));
-  RETURN_NOT_OK(AppendPySequence(seq, size, real_type, builder.get(), from_pandas));
-  return builder->Finish(out);
-}
+  // Create the sequence converter, initialize with the builder
+  std::unique_ptr<SeqConverter> converter;
+  RETURN_NOT_OK(
+      GetConverter(real_type, options.from_pandas, strict_conversions, &converter));
+
+  // Create ArrayBuilder for type, then pass into the SeqConverter
+  // instance. The reason this is created here rather than in GetConverter is
+  // because of nested types (child SeqConverter objects need the child
+  // builders created by MakeBuilder)
+  std::unique_ptr<ArrayBuilder> type_builder;
+  RETURN_NOT_OK(MakeBuilder(options.pool, real_type, &type_builder));
+  RETURN_NOT_OK(converter->Init(type_builder.get()));
+
+  // Convert values
+  if (mask != nullptr && mask != Py_None) {
+    RETURN_NOT_OK(converter->AppendMultipleMasked(seq, mask, size));
+  } else {
+    RETURN_NOT_OK(converter->AppendMultiple(seq, size));
+  }
 
-Status ConvertPySequence(PyObject* obj, MemoryPool* pool, bool from_pandas,
-                         std::shared_ptr<Array>* out) {
-  return ConvertPySequenceReal(obj, -1, nullptr, pool, from_pandas, out);
-}
+  // Retrieve result. Conversion may yield one or more array values
+  std::vector<std::shared_ptr<Array>> chunks;
+  RETURN_NOT_OK(converter->GetResult(&chunks));
+
+  // TODO(wesm): ARROW-2814 comment. Is further casting ever necessary?
```

Comment:
I'm not sure what this comment alludes to? Which casting would we desire here?

Reply:
There was a casting path in the old code. I left this here to remind me; not needed anymore